Sources/TuistEnvKit/HTTP/HTTPClient.swift

```diff
@@ -0,0 +1,123 @@
+        case let .noData(url):
+            return "The request to \(url.absoluteString) returned no data"
+        case let .copyFileError(path, error):
+            return "The file could not be copied into \(path.pathString): \(error.localizedDescription)"
+        case let .missingResource(url):
+            return "Couldn't locate resource downloaded from \(url.absoluteString)"
+        }
+    }
+}
+
+protocol HTTPClienting {
+    /// Fetches the content from the given URL and returns it as a data.
+    ///
+    /// - Parameter url: URL to download the resource from.
+    /// - Returns: Response body as a data.
+    /// - Throws: An error if the request fails.
+    func read(url: URL) throws -> Data
+
+    /// Downloads the resource from the given URL into the passed directory.
+    ///
+    /// - Parameters:
+    ///   - url: URL to download the resource from.
+    ///   - into: Directory where the resource should be placed.
+    /// - Throws: An error if the dowload fails.
+    func download(url: URL, into: AbsolutePath) throws
+}
+
+final class HTTPClient: HTTPClienting {
+    // MARK: - Attributes
+
+    /// URL session.
+    let session: URLSession = .shared
+    let fileHandler: FileHandling = FileHandler.shared
+
+    // MARK: - HTTPClienting
+
+    /// Fetches the content from the given URL and returns it as a data.
+    ///
+    /// - Parameter url: URL to download the resource from.
+    /// - Returns: Response body as a data.
+    /// - Throws: An error if the request fails.
+    func read(url: URL) throws -> Data {
+        var data: Data?
+        var error: Error?
+
+        let semaphore = DispatchSemaphore(value: 0)
+        session.dataTask(with: url) { responseData, _, responseError in
+            data = responseData
+            error = responseError
+            semaphore.signal()
+        }.resume()
+        semaphore.wait()
+
+        if let error = error {
+            throw HTTPClientError.clientError(url, error)
+        }
+        guard let resultData = data else {
+            throw HTTPClientError.noData(url)
+        }
+        return resultData
+    }
+
+    /// Downloads the resource from the given URL into the passed directory.
+    ///
+    /// - Parameters:
+    ///   - url: URL to download the resource from.
+    ///   - into: Directory where the resource should be placed.
+    /// - Throws: An error if the dowload fails.
+    func download(url: URL, into: AbsolutePath) throws {
+        let semaphore = DispatchSemaphore(value: 0)
+        var clientError: HTTPClientError?
+
+        session.downloadTask(with: url) { downloadURL, _, error in
+            defer { semaphore.signal() }
+            if let error = error {
+                clientError = HTTPClientError.clientError(url, error)
+            } else if let downloadURL = downloadURL {
+                let from = AbsolutePath(downloadURL.absoluteString)
+                let to = into.appending(component: from.components.last!)
```

Comment:
minor: I believe `from.basename` achieves the same and is non optional :)

Is the intention to use the temporary name? When I tested this, `from` appears to have a temporary name e.g. `CFNetworkDownload_edu4Be.tmp`.

Reply:
Good catch. We don't want to use the temporary name. I've updated the definition of the method to take the destination path instead of the destination directory. That way is the caller of the function the one that has to indicate the name of the downloaded file.